address: fix memleaks caused from excessive key lengths

xanimo opened the pull request:
This PR includes fixes for #74. Setting to draft until I confirm it fixes the issues in example.c

nooperation commented:
This will resolve that one copy that goes way out of bounds (128 vs 35), but the keys will still have junk at the end when printed because they lack null terminators


generateHDMasterPubKeypair:
```
    if (wif_privkey_master) {
        memcpy_safe(wif_privkey_master, hd_privkey_master, strlen(hd_privkey_master));
    }
    if (p2pkh_pubkey_master) {
        memcpy_safe(p2pkh_pubkey_master, hd_pubkey_master, strlen(hd_pubkey_master));
    }
```

strlen isn't going to include the null terminator. You could technically strlen(...) + 1, but that's extra work and you already know the size of the buffer, so you might as well just copy it all:

```
    if (wif_privkey_master) {
        memcpy_safe(wif_privkey_master, hd_privkey_master, sizeof(hd_privkey_master));
    }
    if (p2pkh_pubkey_master) {
        memcpy_safe(p2pkh_pubkey_master, hd_pubkey_master, sizeof(hd_pubkey_master));
    }
```

The dogecoin_mem_zero calls are also using strlen instead of sizeof at the end of the function. This will still clear the data that needs to be cleared, but it's just extra work and much easier to just zero out the entire buffer

xanimo commented:
> This will resolve that one copy that goes way out of bounds (128 vs 35), but the keys will still have junk at the end when printed because they lack null terminators
> 
> generateHDMasterPubKeypair:
> 
> ```
>     if (wif_privkey_master) {
>         memcpy_safe(wif_privkey_master, hd_privkey_master, strlen(hd_privkey_master));
>     }
>     if (p2pkh_pubkey_master) {
>         memcpy_safe(p2pkh_pubkey_master, hd_pubkey_master, strlen(hd_pubkey_master));
>     }
> ```
> 
> strlen isn't going to include the null terminator. You could technically strlen(...) + 1, but that's extra work and you already know the size of the buffer, so you might as well just copy it all:
> 
> ```
>     if (wif_privkey_master) {
>         memcpy_safe(wif_privkey_master, hd_privkey_master, sizeof(hd_privkey_master));
>     }
>     if (p2pkh_pubkey_master) {
>         memcpy_safe(p2pkh_pubkey_master, hd_pubkey_master, sizeof(hd_pubkey_master));
>     }
> ```
> 
> The dogecoin_mem_zero calls are also using strlen instead of sizeof at the end of the function. This will still clear the data that needs to be cleared, but it's just extra work and much easier to just zero out the entire buffer

Good catch and thank you for the feedback. I cleaned that up and added a couple more commits that are kind of related. I need to open a PR adding missing libs to link for mingw32 before I can really test example.c because that's where I first noticed the extraneous garbage at the end of the addresses. Can't see them anymore but one more step to check and then I'll convert to ready for review.

xanimo commented:
Confirmed these changes fix the amended example.c on x86_64-w64-mingw32:
```
$ wine example.exe 
it looks like wine32 is missing, you should install it.
multiarch needs to be enabled first.  as root, please
execute "dpkg --add-architecture i386 && apt-get update &&
apt-get install wine32"
0048:err:explorer:initialize_display_settings Failed to query current display settings for L"\\\\.\\DISPLAY1".
Mainnet keypair 1:
===============================
Private: QVDzAABp2niSQo73FkERw8YSBV7qTRdPWSt4iApR1sNenNUueaTi
Public:  D6bR7q7Ui8x54q9exJ6pPDdbYGCSZsr4Dt

Mainnet master keypair 2:
===============================
Private: dgpv51eADS3spNJh81YK8cDogdtPfCxZ5skEG8JyFMfb2WbsBfM4rKaeDKV9nv3czcaN8YLeFYnfKvyneumj5GKUgxjTjZzE7W1taX2BnViPjB6
Public:  DAxHCcqeLEX2BnPfQTRgrvYv68LXLqie51

Mainnet master derived keypair 3:
===============================
Private: dgpv51eADS3spNJh81YK8cDogdtPfCxZ5skEG8JyFMfb2WbsBfM4rKaeDKV9nv3czcaN8YLeFYnfKvyneumj5GKUgxjTjZzE7W1taX2BnViPjB6
Public:  DAxHCcqeLEX2BnPfQTRgrvYv68LXLqie51



Keypair (QVDzAABp2niSQo73FkERw8YSBV7qTRdPWSt4iApR1sNenNUueaTi, D6bR7q7Ui8x54q9exJ6pPDdbYGCSZsr4Dt) is valid for mainnet 4.
Keypair (dgpv51eADS3spNJh81YK8cDogdtPfCxZ5skEG8JyFMfb2WbsBfM4rKaeDKV9nv3czcaN8YLeFYnfKvyneumj5GKUgxjTjZzE7W1taX2BnViPjB6, DAxHCcqeLEX2BnPfQTRgrvYv68LXLqie51) is valid for mainnet 5.
Keypair (dgpv51eADS3spNJh81YK8cDogdtPfCxZ5skEG8JyFMfb2WbsBfM4rKaeDKV9nv3czcaN8YLeFYnfKvyneumj5GKUgxjTjZzE7W1taX2BnViPjB6, DAxHCcqeLEX2BnPfQTRgrvYv68LXLqie51) is valid for mainnet 6.


Address D6bR7q7Ui8x54q9exJ6pPDdbYGCSZsr4Dt is valid for mainnet 7.
Address DAxHCcqeLEX2BnPfQTRgrvYv68LXLqie51 is valid for mainnet 8.
Address DAxHCcqeLEX2BnPfQTRgrvYv68LXLqie51 is valid for mainnet 9.


Empty transaction created at index 1.
Input of value 2 dogecoin added to the transaction.
Input of value 10 dogecoin added to the transaction.
Output of value 5 dogecoin added to the transaction.
Change returned to address noxKJyGPugPRN4wqvrwsrtYXuQCk7yQEsy and finalized unsigned transaction saved at index 2.

signature created:
signature compact: 90bddac300243d16dca5e38ab6c80d5848e0d710d77702223bacd6682654f6fe1b5c2e8b1143d8a807d604dc18068b4278facce561c302b0c66a4f2a5a4aa66f
signature DER (+hashtype): 304502210090bddac300243d16dca5e38ab6c80d5848e0d710d77702223bacd6682654f6fe02201b5c2e8b1143d8a807d604dc18068b4278facce561c302b0c66a4f2a5a4aa66f01
signed TX: 0100000002746007aed61e8531faba1af6610f10a5422c70a2a7eb6ffb51cb7a7b7b5e45b4010000006b48304502210090bddac300243d16dca5e38ab6c80d5848e0d710d77702223bacd6682654f6fe02201b5c2e8b1143d8a807d604dc18068b4278facce561c302b0c66a4f2a5a4aa66f00

signature created:
signature compact: 0e19c2a66846109aaae4d29376040fc4f7af1a519156fe8da543dc6f03bb50a13a27495aba9eead2f154e44c25b52ccbbedef084f0caf1deedaca87efd77e4e7
signature DER (+hashtype): 304402200e19c2a66846109aaae4d29376040fc4f7af1a519156fe8da543dc6f03bb50a102203a27495aba9eead2f154e44c25b52ccbbedef084f0caf1deedaca87efd77e4e701
signed TX: 0100000002746007aed61e8531faba1af6610f10a5422c70a2a7eb6ffb51cb7a7b7b5e45b4010000006b48304502210090bddac300243d16dca5e38ab6c80d5848e0d710d77702223bacd6682654f6fe02201b5c2e8b1143d8a807d604dc18068b4278facce561c302b0c66a4f2a5a4aa66f00

All transaction inputs signed successfully.
Final transaction hex: 0100000002746007aed61e8531faba1af6610f10a5422c70a2a7eb6ffb51cb7a7b7b5e45b4010000006b48304502210090bddac300243d16dca5e38ab6c80d5848e0d710d77702223bacd6682654f6fe02201b5c2e8b1143d8a807d604dc18068b4278facce561c302b0c66a40
```
and on i686-w64-mingw32:
```
$ wine example.exe
0048:err:explorer:initialize_display_settings Failed to query current display settings for L"\\\\.\\DISPLAY1".
Mainnet keypair 1:
===============================
Private: QRwe55xsShyaKCa9emDo7XnL4nRZYffTFpDEVXQcry861EVMd4Pk
Public:  DGDE7x4byTAYAr4TcBhNVpgc4ALj8t8eFi

Mainnet master keypair 2:
===============================
Private: dgpv51eADS3spNJh8GaQqbfpgGTER8ptEpyD3WFHKuAhxmLvXaJRozkRmwjBAhVdVxUi2ApAsSZGxLJxpvbhGaAWWkGGz4QoGSzyZBSP38EAfCL
Public:  DQTNfvyrwRALZtfGY91MzadFfc8EiydQb8

Mainnet master derived keypair 3:
===============================
Private: dgpv51eADS3spNJh8GaQqbfpgGTER8ptEpyD3WFHKuAhxmLvXaJRozkRmwjBAhVdVxUi2ApAsSZGxLJxpvbhGaAWWkGGz4QoGSzyZBSP38EAfCL
Public:  DQTNfvyrwRALZtfGY91MzadFfc8EiydQb8



Keypair (QRwe55xsShyaKCa9emDo7XnL4nRZYffTFpDEVXQcry861EVMd4Pk, DGDE7x4byTAYAr4TcBhNVpgc4ALj8t8eFi) is valid for mainnet 4.
Keypair (dgpv51eADS3spNJh8GaQqbfpgGTER8ptEpyD3WFHKuAhxmLvXaJRozkRmwjBAhVdVxUi2ApAsSZGxLJxpvbhGaAWWkGGz4QoGSzyZBSP38EAfCL, DQTNfvyrwRALZtfGY91MzadFfc8EiydQb8) is valid for mainnet 5.
Keypair (dgpv51eADS3spNJh8GaQqbfpgGTER8ptEpyD3WFHKuAhxmLvXaJRozkRmwjBAhVdVxUi2ApAsSZGxLJxpvbhGaAWWkGGz4QoGSzyZBSP38EAfCL, DQTNfvyrwRALZtfGY91MzadFfc8EiydQb8) is valid for mainnet 6.


Address DGDE7x4byTAYAr4TcBhNVpgc4ALj8t8eFi is valid for mainnet 7.
Address DQTNfvyrwRALZtfGY91MzadFfc8EiydQb8 is valid for mainnet 8.
Address DQTNfvyrwRALZtfGY91MzadFfc8EiydQb8 is valid for mainnet 9.


Empty transaction created at index 1.
Input of value 2 dogecoin added to the transaction.
Input of value 10 dogecoin added to the transaction.
Output of value 5 dogecoin added to the transaction.
Change returned to address noxKJyGPugPRN4wqvrwsrtYXuQCk7yQEsy and finalized unsigned transaction saved at index 2.

signature created:
signature compact: 90bddac300243d16dca5e38ab6c80d5848e0d710d77702223bacd6682654f6fe1b5c2e8b1143d8a807d604dc18068b4278facce561c302b0c66a4f2a5a4aa66f
signature DER (+hashtype): 304502210090bddac300243d16dca5e38ab6c80d5848e0d710d77702223bacd6682654f6fe02201b5c2e8b1143d8a807d604dc18068b4278facce561c302b0c66a4f2a5a4aa66f01
signed TX: 0100000002746007aed61e8531faba1af6610f10a5422c70a2a7eb6ffb51cb7a7b7b5e45b4010000006b48304502210090bddac300243d16dca5e38ab6c80d5848e0d710d77702223bacd6682654f6fe02201b5c2e8b1143d8a807d604dc18068b4278facce561c302b0c66a4f2a5a4aa66f00

signature created:
signature compact: 0e19c2a66846109aaae4d29376040fc4f7af1a519156fe8da543dc6f03bb50a13a27495aba9eead2f154e44c25b52ccbbedef084f0caf1deedaca87efd77e4e7
signature DER (+hashtype): 304402200e19c2a66846109aaae4d29376040fc4f7af1a519156fe8da543dc6f03bb50a102203a27495aba9eead2f154e44c25b52ccbbedef084f0caf1deedaca87efd77e4e701
signed TX: 0100000002746007aed61e8531faba1af6610f10a5422c70a2a7eb6ffb51cb7a7b7b5e45b4010000006b48304502210090bddac300243d16dca5e38ab6c80d5848e0d710d77702223bacd6682654f6fe02201b5c2e8b1143d8a807d604dc18068b4278facce561c302b0c66a4f2a5a4aa66f00

All transaction inputs signed successfully.
Final transaction hex: 0100000002746007aed61e8531faba1af6610f10a5422c70a2a7eb6ffb51cb7a7b7b5e45b4010000006b48304502210090bddac300243d16dca5e38ab6c80d5848e0d710d77702223bacd6682654f6fe02201b5c2e8b1143d8a807d604dc18068b4278facce561c302b0c66a40
```
Opening up as ready for review.

michilumin reviewed (approved):
woo ok good, fixes those example.c errors that have been bugging for a while.. new values look right to me.

this should close #74
